### PR TITLE
etcdserver: configurable backend size quota

### LINKED ISSUE #4894: make maxdb size configurable
I would suggest to configure it as 2GB by default and 8GB as max (internally we have 10GB limitation).

We can extend the limitation once we feel more confident about larger data set.

### PR DESCRIPTION
**Pull request by heyitsanthony**
Configurable with the flag --experimental-quota-backend-bytes and
through ServerConfig.QuotaBackendBytes.

Fixes #4894

### CODE REVIEW

File: etcdmain/config.go [hunk 1]
```diff
@@ -224,6 +225,7 @@ func NewConfig() *config {
 
 	// demo flag
 	fs.IntVar(&cfg.autoCompactionRetention, "experimental-auto-compaction-retention", 0, "Auto compaction retention in hour. 0 means disable auto compaction.")
+	fs.Int64Var(&cfg.quotaBackendBytes, "experimental-quota-backend-bytes", 0, "Raise alarms when backend size exceeds the given quota. 0 means use default quota.")
```
Comment (xiang90) [inline etcdmain/config.go]: remove experimental? we are already on 3.0 branch. i will remove the other one later.

File: storage/backend/backend.go [hunk 2]
```diff
@@ -41,6 +41,15 @@ var (
 	InitialMmapSize = int64(10 * 1024 * 1024 * 1024)
 )
 
+const (
+	// DefaultQuotaBytes is the number of bytes the backend Size may
+	// consume before exceeding the space quota.
+	DefaultQuotaBytes = int64(2 * 1024 * 1024 * 1024)
```
Comment (xiang90) [inline storage/backend/backend.go]: comment the number? // 2GB

File: storage/backend/backend.go [hunk 3]
```diff
@@ -41,6 +41,15 @@ var (
...
+	// MaxQuotaBytes is the maximum number of bytes suggested for a backend
+	// quota. A larger quota may lead to degraded performance.
+	MaxQuotaBytes = int64(8 * 1024 * 1024 * 1024)
```
Comment (xiang90) [inline storage/backend/backend.go]: comment the number? // 8GB

File: etcdserver/quota.go [hunk 4]
```diff
@@ -44,7 +50,19 @@ const (
...
+		// disable quotas if negative
+		plog.Warningf("disabling backend quota")
+		return &passthroughQuota{}
+	}
+	if s.cfg.QuotaBackendBytes == 0 {
+		// use default size if no quota size given
+		return &backendQuota{s, backend.DefaultQuotaBytes}
+	}
+	if s.cfg.QuotaBackendBytes > backend.MaxQuotaBytes {
+		plog.Warningf("backend quota %v exceeds recommended maximum quota %v", s.cfg.QuotaBackendBytes, backend.MaxQuotaBytes)
```
Comment (xiang90) [inline etcdserver/quota.go]: probably we should just set it back to the recommended max one...

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Lgtm.

**Pull request merged**